src/coreclr/src/vm/jithelpers.cpp

```diff
@@ -5233,6 +5234,73 @@ void JIT_Patchpoint(int* counter, int ilOffset)
 
 #endif // FEATURE_ON_STACK_REPLACEMENT
 
+HCIMPL2(void, JIT_ClassProfile, Object *obj, void* tableAddress)
+{
+    FCALL_CONTRACT;
+    FC_GC_POLL_NOT_NEEDED();
+
+    OBJECTREF objRef = ObjectToOBJECTREF(obj);
+    VALIDATEOBJECTREF(objRef);
+
+    ICorJitInfo::ClassProfile* const classProfile = (ICorJitInfo::ClassProfile*) tableAddress;
+    const int count = classProfile->Count++;
```

Comment:
Does this need to use volatile and/or interlocked operations?

In `!_DEBUG`, it is a valid C/C++ optimization to get rid of the `count` local and re-fetch it below, ie transform the line below to `classProfile->ClassTable[classProfile->Count]`. I would potentially lead to buffer overruns or skipped slots in `ClassProfile`

Reply:
Yes it should be volatile so it's just read once. Likewise for `s_rng`.

It is OK to drop updates because of races, and we're trying to keep overhead to a minimum, so no need to interlock.